[Concurrency] Polish up the new custom executor interface header.

al45tair opened the pull request:
Added many more comments to the header; also tweaked a couple of things relating to the time-based enqueuing functions.

rdar://137885540

al45tair committed:
[Concurrency] Polish up the new custom executor interface header.

Added many more comments to the header; also tweaked a couple of things
relating to the time-based enqueuing functions.

rdar://137885540

al45tair commented:
@swift-ci Please smoke test

al45tair commented:
(This is based on feedback I've received on the `ExecutorImpl.h` header. There may be further feedback yet, so I'm going to keep this open until everyone concerned is happy.)

kubamracek commented on stdlib/public/Concurrency/ExecutorImpl.h:
```diff
@@ -118,7 +111,11 @@ static inline SwiftJobPriority swift_job_getPriority(SwiftJob *job) {
+/// task is cleaned up.
+///
+/// N.B. **Requires that the job kind is SwiftJobKindTask.**
 void *swift_job_alloc(SwiftJob *job, size_t size);
```
There is also a "stack discipline" requirement when using swift_job_alloc+swift_job_dealloc, correct? If yes, sound like it would be good to document in a comment, too.

al45tair replied:
Good point, we should note that fact.

al45tair committed:
[Concurrency] Add a note about the stack discipline on the allocator.

We should mention that deallocation can only proceed in the reverse
order to the allocation order.

rdar://137885540

al45tair commented:
@swift-ci Please smoke test

ktoso commented on stdlib/public/Concurrency/ExecutorImpl.h:
```diff
@@ -240,57 +242,207 @@ enum {
 // -- Functions that executors must implement ----------------------------------
 
+// Your executor MUST provide definitions of ALL of these functions,
+// however some implementations may be trivial or just assert.
```
👍

ktoso commented on stdlib/public/Concurrency/ExecutorImpl.h:
```diff
@@ -240,57 +242,207 @@ enum {
-/// Drain the main executor's queue, processing jobs enqueued on it; this
-/// should never return.
+/// Drain the global executor's queue, processing jobs enqueued on it; this
+/// should never return.  This is called by the async main entry point.
```
👍

ktoso commented on stdlib/public/Concurrency/DispatchGlobalExecutor.cpp:
```diff
@@ -56,6 +56,15 @@
+  SwiftJobDispatchHasLongObjectHeader = sizeof(void *) == sizeof(int),
+  SwiftJobDispatchLinkageIndex = SwiftJobDispatchHasLongObjectHeader ? 1 : 0,
+  SwiftJobDispatchQueueIndex = SwiftJobDispatchHasLongObjectHeader? 0 : 1
+};
```
👍

ktoso reviewed (approved):
AFAICS looks good, nothing too shocking in here

ktoso commented on stdlib/public/Concurrency/ExecutorImpl.h:
```diff
@@ -118,10 +111,22 @@ static inline SwiftJobPriority swift_job_getPriority(SwiftJob *job) {
-/// Allocate memory associated with a job.
+/// Allocate memory associated with a job.  Memory is allocated using a
+/// per-task allocator and will be disposed of automatically when the
+/// task is cleaned up.
```
maybe point out here rather than just in the `swift_job_dealloc` about the stack discipline?

Because this doc does not allude to the fact that I _have to_ figure out where to dealloc so we get the right order in many cases.

al45tair replied:
Hmmm. I suppose. I was thinking someone would be looking directly at the file, but perhaps they're using an editor like Xcode that's picking up the doc comments.

al45tair commented:
@swift-ci Please smoke test

al45tair commented:
@swift-ci Please smoke test Linux platform

al45tair commented:
@swift-ci Please smoke test Linux platform

al45tair committed:
[Concurrency] Fix a build error.

`long long` isn't directly compatible with `int64_t` or `uint64_t`.

rdar://137885540

al45tair commented:
@swift-ci Please smoke test

al45tair commented:
@swift-ci Please smoke test

al45tair commented:
@swift-ci Please smoke test

al45tair commented:
@swift-ci Please test

al45tair commented:
We aren't going to merge this, I don't believe.
